pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,232 @@
+		shutdownGracePeriodRequested:    shutdownGracePeriodRequested,
+		shutdownGracePeriodCriticalPods: shutdownGracePeriodCriticalPods,
+		clock:                           clock.RealClock{},
+	}
+}
+
+// Start starts the node shutdown manager and will start watching the node for shutdown events.
+func (m *Manager) Start() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+	if m.shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+
+	systemBus, err := getSystemDbus()
+	if err != nil {
+		return err
+	}
+	m.dbusCon = systemBus
+
+	currentInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+	if err != nil {
+		return err
+	}
+
+	// If the logind's InhibitDelayMaxUSec as configured in (logind.conf) is less than shutdownGracePeriodRequested, attempt to update the value to shutdownGracePeriodRequested.
+	if m.shutdownGracePeriodRequested > currentInhibitDelay {
+		err := m.dbusCon.OverrideInhibitDelay(m.shutdownGracePeriodRequested)
+		if err != nil {
+			return err
+		}
+
+		err = m.dbusCon.ReloadLogindConf()
+		if err != nil {
+			return err
+		}
+
+		// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
+		updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+		if err != nil {
+			return err
+		}
+
+		if updatedInhibitDelay != m.shutdownGracePeriodRequested {
+			return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
+		}
+	}
+
+	err = m.aquireInhibitLock()
+	if err != nil {
+		return err
+	}
+
+	events, err := m.dbusCon.MonitorShutdown()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case isShuttingDown := <-events:
+				klog.V(1).Infof("Shutdown manager detected new shutdown event, isNodeShuttingDownNow: %t", isShuttingDown)
+				m.nodeShuttingDownMutex.Lock()
+				m.nodeShuttingDownNow = isShuttingDown
+				m.nodeShuttingDownMutex.Unlock()
+				if isShuttingDown {
+					m.processShutdownEvent()
+				} else {
+					m.aquireInhibitLock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *Manager) aquireInhibitLock() error {
+	lock, err := m.dbusCon.InhibitShutdown()
+	if err != nil {
+		return err
+	}
+	m.inhibitLock = lock
+	return nil
+}
+
+// ShutdownStatus will return an error if the node is currently shutting down.
+func (m *Manager) ShutdownStatus() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+
+	m.nodeShuttingDownMutex.Lock()
+	defer m.nodeShuttingDownMutex.Unlock()
+
+	if m.nodeShuttingDownNow {
+		return fmt.Errorf("node is shutting down")
+	}
+	return nil
+}
+
+func (m *Manager) processShutdownEvent() error {
+	klog.V(1).Infof("Shutdown manager processing shutdown event")
```

Comment:
should this be a proper event, not just a log message?

Reply:
Hmm, I'm not sure. There will already be an event emitted as part of the node going to not ready here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L575

Reply:
makes sense